src/symbolic_choice.ml

```diff
@@ -319,128 +217,198 @@ module Multicore = struct
-    | Trap t -> WQ.push (ETrap t, thread) st.global.r
-    | Assert c ->
-      if check c thread st.solver then cont.k thread (E_st st) ()
-      else
-        let no = S.Bool.not c in
-        let thread = { thread with pc = no :: thread.pc } in
-        WQ.push (EAssert c, thread) st.global.r
-    | Bind (v, f) ->
-      let k thread (E_st st) v =
-        let r = f v in
-        local_push st (H (thread, r, cont))
-      in
-      step thread v { k } st
-    | Choice cond ->
-      let cases = list_select cond thread st.solver in
-      List.iter (fun (case, thread) -> cont.k thread (E_st st) case) cases
-    | Choice_i32 i ->
-      let cases = list_select_i32 i thread st.solver in
-      List.iter (fun (case, thread) -> cont.k thread (E_st st) case) cases
+    | Stop -> ()
+    | Now x -> WQ.push x w
+    | Yield (_prio, f) -> WQ.push f q
+    | Choice (m1, m2) ->
+      handle_aux m1 q w;
+      handle_aux m2 q w
+
+  let rec work solver (q : 'a work_queue) (w : 'a res_que) =
+    match WQ.take q true with
+    | None -> ()
+    | Some f -> begin
+      handle_aux (f solver) q w;
+      WQ.end_pledge q;
+      work solver q w
+    end
 
   let init_global () =
     let w = WQ.init () in
     let r = WQ.init () in
-    let start_counter = Counter.create () in
-    { w; r; start_counter }
+    { w; r }
 
   let push_first_work g thread t =
-    let k thread _st v = WQ.push (EVal v, thread) g.r in
-    WQ.push (H (thread, t, { k })) g.w
-
-  let spawn_producer global _i =
-    let solver = fresh_solver () in
-    let st = { solver; next = None; global } in
-    let rec producer (H (thread, t, cont)) =
-      step thread t cont st;
-      match st.next with
-      | Some h ->
-        st.next <- None;
-        producer h
-      | None -> ()
-    in
+    let first = run t thread in
+    WQ.push first g.w
+
+  let spawn_producer global =
+    WQ.make_pledge global.r;
     Domain.spawn (fun () ->
+        let solver = fresh_solver () in
         try
-          WQ.with_produce global.r
-            ~started:(fun () -> Counter.incr global.start_counter)
-            (fun () -> WQ.produce global.w producer)
+          work solver global.w global.r;
+          WQ.end_pledge global.r
         with e ->
           let bt = Printexc.get_raw_backtrace () in
           WQ.fail global.w;
           Printexc.raise_with_backtrace e bt )
 
-  let rec loop_and_do (s : 'a Seq.t) f : 'a Seq.t =
-   fun () ->
-    match s () with
-    | Cons (s, t) -> Cons (s, loop_and_do t f)
-    | Nil ->
-      f ();
-      Nil
+  type 'a run_result = ('a eval * Thread.t) Seq.t
 
   let run ~workers t thread =
     let global = init_global () in
     push_first_work global thread t;
-    let producers = Array.init workers (spawn_producer global) in
-    Counter.wait_n global.start_counter workers;
-    loop_and_do (WQ.read_as_seq global.r) (fun () ->
-        Array.iter Domain.join producers )
+    for _ = 0 to workers do
+      let _ = spawn_producer global in
```

Comment:
do not use `let _ =`, if you change the type of `spawn_producer`, you will do a partial application without noticing and this is usually hard to debug, adding a type annotation is enough to fix the potential issue